src/datafile.rs

```diff
@@ -95,23 +95,14 @@ impl DataFile {
     }
 
     pub fn write(&mut self, key: &[u8], value: &[u8], timestamp: u128) -> ErrorResult<u64> {
-        use std::io::Write;
-
         let entry = Entry {
             timestamp,
             key: key.to_vec(),
             value: value.to_vec(),
         };
 
         let offset = self.file.seek(SeekFrom::Current(0))?;
-
-        let encoded: Vec<u8> = bincode::serialize(&entry)?;
-
-        let written = self.file.write(&encoded);
-        if let Err(err_msg) = written {
-            return Err(Box::new(err_msg));
-        }
-
+        bincode::serialize_into(&mut *self.file, &entry)?;
```

Comment:
Most serde libraries let you serialize and deserialize directly from `std::io::Read` and `std::io::Write` implementations.

So, we can skip serializing to heap just to then write it out to the file by directly serializing to the file.

The `&mut *self.file` is needed because of how you have `DerefMut` implemented.. (more on that later)